File: stara/2017/c++/zad4_2.cpp
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
using namespace std;

//funkcja zwracajaca wartosc w systemie dziesietnym danego zapisu binarnego
int toDecimal1(string line)
{
    int output = 0;

    for (int i = line.size() - 1; i >= 0; i--)
    {
        if (line[line.size() - i - 1] == '1')
            output += pow(2, i);
    }

    return output;
}

//funkcja sprawdzajaca czy string jest poprawnie zapisany
bool isCorrect(string line)
{
    //sprawdzanie czy napis sklada sie z cztero-znakowych elementow
    if (line.size() % 4 == 0)
    {
        string readyToCheck = "";

        //przechodzenie po napisie
        for (int i = 0; i < line.size(); i++)
        {
            if ((i + 1) % 4 == 0)
            {
                readyToCheck += line[i];

                //sprawdzanie czy wartosc nie przekracza 9
                if (toDecimal1(readyToCheck) > 9)
                    return false;

                readyToCheck = "";
            }
            else
                readyToCheck += line[i];
        }

        return true;
    }
    else
        return false;
}

string zad4_2()
{
    string line;
    vector<string> content;

    //wczytywanie danych z pliku
    fstream file("../dane/binarne.txt");

    if (file.is_open())
    {
        while (getline(file, line))
            content.push_back(line);
    }
    file.close();

    //sprawdzanie ile jest niepoprawnie zapisanych napisow oraz wczytywanie do osobnego vectora
    int howManyFalse = 0;
    vector<string> falseNumbers;

    for (int i = 0; i < content.size(); i++)
    {
        if (isCorrect(content[i]) == false)
        {
            howManyFalse++;
            falseNumbers.push_back(content[i]);
        }
    }

    //sprawdzanie ile wynosi najkrotszy napis sposrod niepoprawnych
    int minLength = 10000;

    for (int i = 0; i < falseNumbers.size(); i++)
    {
        if (falseNumbers[i].size() < minLength)
            minLength = falseNumbers[i].size();
    }

    return "4.2. Liczba slow niepoprawnych: " + to_string(howManyFalse) + "; Dlugosc najkrotszego niepoprawnie zapisanego napisu: " + to_string(minLength);
}

File: stara/2017/python/zad4_2.py
# wczytanie napisow
with open('../dane/binarne.txt') as f:
    data = []
    for word in f.readlines():
        data.append(word[:-1])

# licznik
badtens = 0
# najkrotsze
shortest = '1'*32
# przejscie po kazdym napisie
for word in data:
    # przejscie po kazdym segmentcie napisu
    for i in range(len(word)//4):
        segment = word[i*4:(i+1)*4]
        # jezeli segment w dziesietnym jest wieksze niz 9 to zlicz
        if int(segment, base=2) > 9:
            badtens += 1
            # jezeli jest owy napis krotszy, zapisz
            if len(word) < len(shortest):
                shortest = word
            break


# wyswietlenie odpowiedzi
answer = f'4.2. Liczba napisow niepoprawnie zapisanych w BCD: {badtens}; Dlugosc najkrotszego takiego napisu: {len(shortest)}'
print(answer)
